### PR TITLE
[GLUTEN-3582][CORE][VL][CH] Refactor filter pushdown logic

### PR DESCRIPTION
**Pull request by baibaichen**
## What changes were proposed in this pull request?

Vanilla spark just push down part of filter condition into scan, however gluten can push down all filters in some cases, it was done in #4132. But, Clickhouse backend can not do this in parquet file format. we can add this later

This PR add `postProcessPushDownFilter` in `SparkPlanExecApi`,  `CHSparkPlanExecApi` overwrite this funtion to add its own logic.

### Note for CH
Consider TPCH 22 `c_acctbal > (select avg(c_acctbal) from customer where ...)`.   it can be push down before this PR.  I disable it for ch backend, since I want to make push down functionality same as vanilla spark first.

## Update for `ScanTransformerFactory`
Before this PR, `createxxxTransformer` assume pass the extra ilters, after this PR, we pass all push down filters, this is because we may remove unsupported filters  

```scala 
def createFileSourceScanTransformer(
      scanExec: FileSourceScanExec,
      reuseSubquery: Boolean,
      allPushDownFilters: Option[Seq[Expression]] = None,  // extraFilters: Seq[Expression] = Seq.empty,
      validation: Boolean = false): FileSourceScanExecTransformer = {}

  def createBatchScanTransformer(
      batchScan: BatchScanExec,
      reuseSubquery: Boolean,
      allPushDownFilters: Option[Seq[Expression]] = None,  //pushdownFilters: Seq[Expression] = Seq.empty,
      validation: Boolean = false): SparkPlan = {}
```

## How was this patch tested?
Using existed UT

### CODE REVIEW

**Comment by baibaichen** [comment]
@liujiayi771 please review it.

**Comment by zhztheplayer** [comment]
Changes to common module looks good to me. Thanks.

File: gluten-core/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala [hunk 1]
```diff
@@ -384,30 +384,27 @@ object FilterHandler extends PredicateHelper {
     (ExpressionSet(filters) -- ExpressionSet(scanFilters)).toSeq
 
   // Separate and compare the filter conditions in Scan and Filter.
-  // Push down the remaining conditions in Filter into Scan.
+  // Try push down the remaining conditions in Filter into Scan.
```
Comment (liujiayi771) [inline gluten-core/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala:387]: nit: Try to push down.

**Comment by liujiayi771** [comment]
LGTM. Thanks.

File: gluten-core/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala (hunk 1 above)
Comment (baibaichen) [reply gluten-core/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala:387]: Let's fix in the follow up PR

**Review by zhouyuan — APPROVED** [review]
👍

**Comment by abbas-gadhia** [comment]
Hello @baibaichen 
you mentioned above that "however gluten can push down all filters in some cases". How is this possible when the underlying parquest scan does not support filters such as `concat(name, surname) = "XY"` or `abs(profit) = 20`

**Pull request merged**